### PR TITLE
UI: Add (eRTMP) multitrack video output

### PR DESCRIPTION
**Pull request by palana**
~~Note that this PR includes a bunch of merges from currently open PRs and a FIXME commit to help with testing, which will go away as soon as the dependency PRs are merged~~

### Description
This implements multitrack video support for Twitch Enhanced Broadcasting and Amazon IVS Multitrack Video

### Motivation and Context
Multitrack video/Enhanced Broadcasting allows sending multiple video tracks/renditions from the users machine, resulting in generally higher quality renditions at lower bitrates compared to single track streaming with server side transcoding, since e.g. losses roundtripping frames through another decode/encode step are removed.
Enhanced Broadcasting/multitrack video uses server side configuration to provide easier onboarding for users, and balanced values for the number of renditions and their bitrates.

### How Has This Been Tested?
This has been tested in the Twitch Enhanced Broadcasting beta

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: UI/window-basic-settings-stream.cpp
Comment (tytan652) [inline UI/window-basic-settings-stream.cpp:662]: Turn `UpdateMultitrackVideo` into a slot and connect it to the widget.

File: UI/window-basic-settings-stream.cpp
Comment (tytan652) [inline UI/window-basic-settings-stream.cpp:663]: Turn `UpdateMultitrackVideo` into a slot and connect to widgets.

File: UI/goliveapi-postdata.cpp
Comment (tytan652) [inline UI/goliveapi-postdata.cpp:22]: The user must be able to reject having their system information being sent to a third party.

File: UI/obs-frontend-api/obs-frontend-api.h
Comment (tytan652) [inline UI/obs-frontend-api/obs-frontend-api.h:66]: With how messy service and streaming output codebase is becoming, I would be heavily against adding any Frontend API unless it's for untangling the whole broadcast code and the whole "golive" stuff and put most of it in the service code.

File: UI/CMakeLists.txt
Comment (tytan652) [inline UI/CMakeLists.txt:95]: "goliveapi" is clearly Twitch/Amazon made thing. I would really prefer that most of this goes to where it belong: in the service plugin side.

We already have the Broadcast feature from YT to untangle to a proper service with an agnostic API to manage broadcast, this would worsen the situation.

**Review by tytan652 — CHANGES_REQUESTED** [review]
This feature is clearly not implemented in a future-proof way, it tangles even more service behavior in the UI that should be plugin-ified (with APIs).

Continuously shoehorning service integrations was already a mistake, and this PR goes to the same wrong direction rather than trying to go to a better direction.

File: UI/goliveapi-postdata.cpp
Comment (koolscooby) [reply UI/goliveapi-postdata.cpp:22]: > The user must be able to reject having their system information being sent to a third party.

The user can reject having their system information being sent by not enabling Enhanced Broadcasting or Multitrack Video in the UI.

In other words: this information is never sent to the destination service if the "Enable Enhanced Broadcasting" or "Enable Multitrack Video" are unchecked. In the UI, the user is presented with a checkbox to enable Enhanced Broadcasting or Multitrack Video.

If the feature is enabled via the UI checkbox, then `GetClientCapabilities` is required to fetch the configuration settings and further, the `GetClientCapabilities` API requires this information to vend the best possible, safe ladder for every setup.

Comment (tytan652) [reply UI/goliveapi-postdata.cpp:22]: This is bad, the user needs be aware of this and make an explicit choice to send their data.
Hiding this detail is really bad, IIRC the data that are sent can be assigned as PII.

Comment (Fenrirthviti) [reply UI/goliveapi-postdata.cpp:22]: The user is made aware of this with an explicit opt-in and notification when they enable the feature. Nothing is hidden, and I'm not sure why you're assuming it is.

Comment (tytan652) [reply UI/goliveapi-postdata.cpp:22]: I didn't see the notification part when the check box is toggled in settings, maybe I'm looking at the wrong place.

Sorry then.

Comment (tytan652) [reply UI/goliveapi-postdata.cpp:22]: Found it, it's a label. I'm sorry.

To me, it looks like a bad pattern to block a feature based on if the user wants to share data or not.
But this is out of scope.

This can be put as resolved.

File: UI/window-basic-settings-stream.cpp
Comment (palana) [reply UI/window-basic-settings-stream.cpp]: changed right after https://github.com/obsproject/obs-studio/pull/10633#event-12771455631

File: UI/obs-frontend-api/obs-frontend-api.h
Comment (palana) [reply UI/obs-frontend-api/obs-frontend-api.h:66]: since there's currently nothing using this event it should be ok to just remove it; let me know if you want me to do that

File: UI/window-basic-settings-stream.cpp
Comment (tytan652) [reply UI/window-basic-settings-stream.cpp]: Can be put as resolved.

File: UI/obs-frontend-api/obs-frontend-api.h
Comment (palana) [reply UI/obs-frontend-api/obs-frontend-api.h]: Removed in a fixup commit after https://github.com/obsproject/obs-studio/pull/10633#event-12771825219

Comment (RytoEX) [reply UI/obs-frontend-api/obs-frontend-api.h]: Resolving.

**Comment by RytoEX** [comment]
Rebased first. Now pushing cleanups:
1. The multitrack audio commits don't actually need to be in here because the later commits don't touch the same code. VOD Track just won't work without them, but removing the commits reduces some noise here.
2. We can drop "UI: Keep weak stream output ref in status bar" since it was handled in #10680.

**Comment by palana** [comment]
@RytoEX, @Warchamp7: moved the remaining service settings into groupboxes after <https://github.com/obsproject/obs-studio/pull/10633#event-12870619591>
image
not super happy with the naming of either of the new groupboxes

**Review by RytoEX — CHANGES_REQUESTED** [review]
Started a review a while back and forgot to submit it. Submitting it now so get this one feedback item raised.

File: UI/qt-helpers.hpp
Comment (palana) [reply UI/qt-helpers.hpp:37]: huh, yea, that was meant to be 6.6.2; will push a fix in a moment

**Comment by palana** [comment]
> I tried to autosquash locally, but it looks like some of the autosquash commits are trying to apply changes to commits before the actual base changes the autosquash commits try to fixup. Could you squash the commits?
> 
> Per @Warchamp7 , the UI itself is okay. @Lain-B and I are still doing code review.

@RytoEX: commits are squashed, apologies for the unclean fixup commits

**Review by Lain-B — APPROVED** [review]
Thanks for rebasing. There are a few minor things, such as Linux/macOS just not being implemented at all and failing with "GPU not supported", but generally it has my approval. We'll probably just have to ifdef out the Linux/macOS codepath for it and disable it for the time being but otherwise again it generally has my approval.

**Comment by RytoEX** [comment]
Rebased to latest master. May push trivial fixes (text nits + minor cleanups or just text nits) before merging.

**Comment by koolscooby** [comment]
> Thanks for rebasing. There are a few minor things, such as Linux/macOS just not being implemented at all and failing with "GPU not supported", but generally it has my approval. We'll probably just have to ifdef out the Linux/macOS codepath for it and disable it for the time being but otherwise again it generally has my approval.

If possible, we'd like to keep the implementation as-is so we can release server-side updates to start vending ladders for Linux, MacOS, AMD AMF, and Intel QSV without requiring additional changes to OBS.

In the short term, we can improve the error message that is shown so that it makes it more apparent to users that the reason it is unsupported is not the build of OBS Studio, but instead the service that vends the configuration.

**Review by RytoEX — COMMENTED** [review]
Leaving a brief review comment for posterity.

**Review by RytoEX — APPROVED** [review]
The CI failures are unrelated. The GitHub Actions Ubuntu runners are restarting due to some issue related to apt packages that indicate they require service restarts.

**Pull request merged**
